explainers/signed-exchange-subresource-subtitution-explainer.md

```diff
@@ -0,0 +1,455 @@
+   ```
+   <link rel="prefetch" href="https://feed.example/sxg.publisher.example/article.html.sxg">
+   ```
+When the UA prefetches the signed exchange (article.html.sxg), the aggregator
+server includes a declaration that one of `article.html`'s subresources
+(https://cdn.publisher.example/lib.js) is also available from the same
+aggregator. The aggregator server expresses this by serving `article.html.sxg`
+with a `Link` header identifying the subresource's alternate form:
+
+```
+Link: <https://feed.example/sxg.publisher.example/lib.js.sxg>;
+        rel="alternate";
+        type="application/signed-exchange;v=b3";
+        anchor="https://cdn.publisher.example/lib.js"
+```
+
+To prevent an attacker from loading an incompatible version of the subresource,
+the resource _inside_ the signed exchange has to identify the exact version of
+the replacement signed exchange using a `Link: ... rel="allowed-alt-sxg"` with
+the hash of the signed headers (which themselves include a hash of the content).
+
+To prevent a tracker from conveying a user ID in their choice of which
+subresources to prefetch, the inner resource has to preload the same
+subresources that the aggregator prefetches.
+And the inner response of the main resource signed exchange (article.html.sxg)
+has a preload header and an allowed-alt-sxg header:
+
+```
+Link: <https://cdn.publisher.example/lib.js>;
+        rel="preload";
+        as="script"
+Link: <https://cdn.publisher.example/lib.js>;
+        rel="allowed-alt-sxg";
+        header-integrity="sha256-XXXXXX"
+```
+
+The UA recursively prefetches lib.js.sxg.
+
+If the user navigates to the expected article, both the main resource of the
+article and the script subresource are loaded from the prefetched signed
+exchanges.
+
+# Proposal
+
+## Privacy-preserving prefetching subresources
+
+1. Introduce two new cache mechanism SignedExchangeCache and
+PrefetchedSignedExchangeCache in a Document.
+   - SignedExchangeCache keeps the signed exchanges which are
+prefetched from the Document.
+   - PrefetchedSignedExchangeCache keeps the signed
+exchanges which were passed from the referrer Document which triggered the
+navigation to the current Document.
+1. While processing
+"[prefetch](https://html.spec.whatwg.org/multipage/links.html#link-type-prefetch)"
+link (eg: `<link rel="prefetch"
+href="https://cdn.feed.example/cdn.publisher.example/lib.js.sxg" as="script">`):
+   - If succeeded prefetching the signed exchange (including merkle integrity
+   check of the body), stores the parsedExchange which is the result of
+   [parsing a signed exchange](https://wicg.github.io/webpackage/loading.html#ref-for-parsing-a-signed-exchange)
+   (= inner request URL and inner response) to the SignedExchangeCache.
+1. While
+[navigating across documents](https://html.spec.whatwg.org/multipage/browsing-the-web.html#navigating-across-documents),
+copy the parsedExchanges in the SignedExchangeCache of the source document to
```

Comment:
Presumably except for the one that serves the navigation itself?

Reply:
Yes.
Added.